### PR TITLE
Solution for format phone number challenge

### LINKED ISSUE #15: Format Phone Number Challenge
## Description

Format Phone Number Challenge
---

## Behavior

Write a function that accepts an array of 10 integers (between 0 and 9) and returns a string formatted as a phone number.

---

#### Input:
format_phone_number([1, 2, 3, 4, 5, 6, 7, 8, 9, 0])
#### Output:
"(123) 456-7890"

#### Input:
format_phone_number[5, 4, 3, 4, 7, 6, 2, 8, 1, 0])
#### Output:
"(543) 476-2810"

### PR DESCRIPTION
**Pull request by KarimMakki**
---
name: Format Phone Number Challenge
about: A template PR for code review with a checklist
---

## Behavior
Returns a string formatted as a phone number from a list of integers

 The function should:
  - Accept exactly 10 integers as input.
  - Ensure all integers are between 0 and 9.
  - Return a string in the format: "(XXX) XXX-XXXX", where X represents the digits from the input array.
  
### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_format_phone_number.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import unittest
+
+from ..format_phone_number import format_phone_number
+
+
+class TestFormatPhoneNumber(unittest.TestCase):
+    """Tests for format phone number function"""
+
+    # Standard test cases
+    def test_format_phone_number_mixed_digits(self):
```
Comment (Ammmoe) [inline solutions/tests/test_format_phone_number.py]: I think the name is a bit redundant because we can just say `test_mixed_digits(self)` instead of repeating the name of the function `format_phone_number`.

File: solutions/tests/test_format_phone_number.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        """It should format mixed digits as a phone number string"""
+        actual = format_phone_number([0, 5, 6, 2, 3, 1, 7, 6, 5, 0])
+        expected = "(056) 231-7650"
+        self.assertEqual(actual, expected)
+
+    def test_format_phone_number_single_digit(self):
```
Comment (Ammmoe) [inline solutions/tests/test_format_phone_number.py]: Same idea with line 25 because we can just say `test_single_digit(self)` or `test_same_digit(self)`.

File: solutions/tests/test_format_phone_number.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        """It should format single digit as a phone number string"""
+        actual = format_phone_number([9, 9, 9, 9, 9, 9, 9, 9, 9, 9])
+        expected = "(999) 999-9999"
+        self.assertEqual(actual, expected)
+
+    def test_format_phone_number_ascending_digits(self):
```
Comment (Ammmoe) [inline solutions/tests/test_format_phone_number.py]: Same idea with line 25 and 31. `test_ascending_digits(self)`.

File: solutions/format_phone_number.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+Module contents:
+    - format_phone_number: returns a string formatted as a phone number from a list of integers
+
+Created on 2025-01-04
+@author: KarimMakki
+"""
+
+
+def format_phone_number(phone_number: list) -> str:
```
Comment (Ammmoe) [inline solutions/format_phone_number.py]: I think it is better to use `...list[int]` instead of just `list` in the type annotation because it's more concise and clearer to understand.

File: solutions/format_phone_number.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    """Returns a string formatted as a phone number from a list of integers
+
+    The function should:
+    - Accept exactly 10 integers as input.
+    - Ensure all integers are between 0 and 9.
+    - Return a string in the format: "(XXX) XXX-XXXX", where X represents the digits from the input array.
+
+    Parameter:
+    phone_number: a list of integers
```
Comment (Ammmoe) [inline solutions/format_phone_number.py]: I think it'd be clearer to also type in the purpose of the argument as well.

For example: `phone_number (list[int]): A list of integers representing digits of a phone number.`

According to the checklist, Evan might also want us to put `assumptions` under the arguments title but personally, I think that would be redundant because it is already described in the function's behaviour.

Please enlighten me what you think about this part of checklist.

File: solutions/format_phone_number.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    Returns -> str: a string formatted as a phone number
+
+    Raises:
+    AssertionError: if the input type is not a list
+    AssertionError: If the input contains non-integer elements.
+    ValueError: if the input list does not contain exactly 10 integers
+    ValueError: if any integer is not in the range 0 to 9
```
Comment (Ammmoe) [inline solutions/format_phone_number.py]: I think it's be clearer if you format it like...
For example:

Raises: AssertionError
- if the input type is not a list
- if the input contains non-integer elements

Raises: ValueError
- if the input list does not contain exactly 10 integers
- if any integer is not in the range of 0 to 9

**Review by Ammmoe — CHANGES_REQUESTED** [review]
Thank you for the great solution and test modules as your code passes all CI checks, functions as it is supposed to, and raises all possible errors. To be honest, my review is just nitpicking at this point.

I'd like to request a change on the type annotation of the function on line 15 of `format_phone_number.py`.

My other comments are just suggestions so you can have a look and share your thoughts in the reply.

File: solutions/format_phone_number.py (hunk 5 above)
Comment (KarimMakki) [reply solutions/format_phone_number.py]: I believe I can definitely add the purpose and also the assumptions as follows:

`phone_number (list[int]): a list of integers between 0 and 9 representing digits of a phone number
`

Thanks for the heads up!

File: solutions/format_phone_number.py (hunk 6 above)
Comment (KarimMakki) [reply solutions/format_phone_number.py]: I will be adding it this way, I think it's more clean and easier to read:

    Raises:
        AssertionError:
            - If the input type is not a list.
            - If the input contains non-integer elements.
        ValueError:
            - If the input list does not contain exactly 10 integers.
            - If any integer is not in the range of 0 to 9.

File: solutions/tests/test_format_phone_number.py (hunk 1 above)
Comment (KarimMakki) [reply solutions/tests/test_format_phone_number.py]: Agree, I will be modifying this as you mentioned

File: solutions/tests/test_format_phone_number.py (hunk 2 above)
Comment (KarimMakki) [reply solutions/tests/test_format_phone_number.py]: Thanks, I will be modifying it, it's redundant as you said

File: solutions/format_phone_number.py (hunk 4 above)
Comment (KarimMakki) [reply solutions/format_phone_number.py]: I agree, will add `list[int]` instead

File: solutions/tests/test_format_phone_number.py (hunk 3 above)
Comment (KarimMakki) [reply solutions/tests/test_format_phone_number.py]: I will be modifying that too

**Comment by KarimMakki** [comment]
Thanks Aung for your review, I've made the modifications as per your review and added my insight in each conversation, let me know if there's other thing to modify.

**Review by Ammmoe — APPROVED** [review]
Thanks @KarimMakki for making the changes and providing your insights as well!

I checked the implementation steps if there are more efficient ways to code but the current implementation is already really good so I have nothing to object. Some minor flaws in the docstring have also been addressed.

So I will close this review and merge the pull request now. Thanks!
